README.md
[npm-url]: https://npmjs.org/package/ease-value
[npm-image]: http://img.shields.io/npm/v/ease-value.svg
[travis-url]: https://travis-ci.org/kasparsz/ease-value
[travis-image]: http://img.shields.io/travis/kasparsz/ease-value.svg

# ease-value
[![NPM version][npm-image]][npm-url] [![Build Status][travis-image]][travis-url]

Value easing component which animates single value continuously.

Unlike standard animations EaseValue doesn't have a duration or fixed "to" value and can have "to" value changing frequently, for example
on mouse movement.

## Install

Install with [npm](https://www.npmjs.com/):

```sh
$ npm install ease-value --save
```

## Example

```js
import EaseValue from 'ease-value';

// Create animation with small 'force' value for smoother / slower animation
const mouseX = new EaseValue({'force': 0.1});
const mouseY = new EaseValue({'force': 0.1});

// On mouse movement update 'mouse' animation target value
document.addEventListener('mousemove', function (event) {
    mouseX.to(event.clientX);
    mouseY.to(event.clientY);
}, false);

// On animation step move demo box. Callback will be called twice per animation frame,
// once for mouseX and once for mouseY
mouseX.on('step', updateDemoBoxPosition);
mouseY.on('step', updateDemoBoxPosition);

function updateDemoBoxPosition (value) {
    document.querySelector('#demo-box').style.transform = `translate(${ mouseX.value }px, ${ mouseY.value }px)`;
});
```

To avoid callback being called more than once per animation frame we can use ```EaseValue.multiple(instances)``` and add listener to it instead.

```js
import EaseValue from 'ease-value';

// Listen to all events 
const mouse  = EaseValue.multiple({
    'x': new EaseValue({'force': 0.1}),
    'y': new EaseValue({'force': 0.1})
});

// On mouse movement update 'mouse' animation target value
document.addEventListener('mousemove', function (event) {
    mouse.to({
        'x': event.clientX,
        'y': event.clientY
    });

    // or you can access each EaseValue individually
    mouse.x.to(event.clientX);
    mouse.y.to(event.clientY);
}, false);

// On animation step move demo box
mouse.on('step', function (value) {
    document.querySelector('#demo-box').style.transform = `translate(${ value.x }px, ${ value.y }px)`;
});
```

## API

### `EaseValue([options])`

#### Options

| Name     | Type    | Usage                                    | Default  |
| -------- | ------- | ---------------------------------------- | -------- |
| ```force```    | Number | Amount of force used for animation / animation speed. Smaller the value, slower the animation will be | ```0.1```     |
| ```precision``` | Number | To prevent unneeded ```step``` event calls and to allow detecting animation end faster, animation value is rounded to this precision. For example with default precision, value between steps will not change by less than 0.01 | ```0.01``` |
| ```easing``` | String | Easing name, currently there is ```easeOut``` and ```linear``` easings | ```"easeOut"``` |
| ```value``` | Number | Initial value, optional. If initial value is not provided, then it will be set when calling ```to``` or ```reset``` for first time | null |

### `on(eventName, callback)`

Adds event listener. To callback is passed current value as first argument.

#### Event names

| Name     | Description  |
| -------- | ------- | ---------------------------------------- | -------- |
| ```start```    | Triggered when animation starts or when animation resumes after previously completing / stoping |
| ```stop```    | Triggered when animation completes / stops |
| ```step```    | Triggered on each animation step after value change, if value didn't changed between steps then 'step' callback won't be called |

### `off(eventName, callback)`

Removes event listener.

### `to(value)`

Set value to which animate to. Calling ```to``` for first time if ```value``` option was not provided is equivalent to calling ```reset```

### `reset(value)`

Set value without using animation, will instantly jump to this value.

### `destroy()`

Removes all listeners, stops animation


## Running tests

Install dev dependencies:

```sh
$ npm install -d && npm test
```

## License

Copyright © 2017, [Kaspars Zuks](https://github.com/kasparsz).
Released under the [MIT license](https://github.com/kasparsz/ease-value/blob/master/LICENSE).

[npm-url]: https://npmjs.org/package/ease-value
[npm-image]: http://img.shields.io/npm/v/ease-value.svg
[travis-url]: https://travis-ci.org/kasparsz/ease-value
[travis-image]: http://img.shields.io/travis/kasparsz/ease-value.svg


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [1.4.1] - 2019-12-18
### Fixed
- Updated dependencies

## [1.4.0] - 2019-12-18
### Added
- Added destroy method to the EaseValue.multiple

## [1.3.0] - 2018-09-20
### Added
- Added EaseValue.multiple
- Fixed too many callback calls 

## [1.2.0] - 2017-04-28
### Added
- ```destroy``` method

## [1.1.2] - 2017-04-13
### Added
- Easing error fix

## [1.1.1] - 2017-04-13
### Added
- Delta JS error fix

## [1.1.0] - 2017-04-13
### Added
- Delta time argument to easing function calls
- Added linear easing function

## [1.0.3] - 2017-04-13
### Fixed
- Fixed animation acceleration if setting new target value in 'stop' event callback

## [1.0.2] - 2017-04-13
### Fixed
- Passing intial value throws JS error

## [1.0.1] - 2017-04-05
### Fixed
- Incorrect reported value when animation ends

## [1.0.0] - 2017-04-04
### Added
- Initial public version


test.js
'use strict';

require('mocha');
var assert = require('assert');

var fs = require('fs');
var EaseValue = require('./dist/ease-value.js');

describe('ease-value', function(done) {



});


src/index.js
/*!
 * ease-value <https://github.com/kasparsz/ease-value>
 *
 * Copyright (c) 2017, Kaspars Zuks.
 * Licensed under the MIT License.
 */


const timing = (function () {
    if (typeof performance !== 'undefined') {
        return performance;
    } else {
        return Date;
    }
})();


/**
 * Event listener mixin with .on, .off and .trigger functionality
 */
class Events {

    constructor () {
        this.listeners = {};
    }

    /**
     * Add event listener
     *
     * @param {string} eventName
     * @param {function} callback
     */
    on (eventName, callback) {
        const listeners = this.listeners;

        if (typeof callback === 'function') {
            listeners[eventName] = listeners[eventName] || [];
            listeners[eventName].push(callback);
        }
    }

    /**
     * Remove event listener
     *
     * @param {string} eventName
     * @param {function} callback
     */
    off (eventName, callback) {
        const callbacks = this.listeners[eventName];
        const index     = callbacks ? callbacks.indexOf(callback) : -1;

        if (index !== -1) {
            callbacks.splice(index, 1);
        }
    }

    /**
     * Trigger event listeners
     *
     * @param {string} eventName
     * @protected
     */
    trigger (eventName, value) {
        const callbacks = this.listeners[eventName];

        for (let i = 0, ii = callbacks ? callbacks.length : 0; i < ii; i++) {
            callbacks[i](value);
        }
    }
}


class EaseValue extends Events {

    static get Defaults () {
        return {
            'value': null,
            'force': EaseValue.defaultForce,
            'precision': EaseValue.defaultPrecision,
            'easing': EaseValue.defaultEasing
        };
    }


    constructor (opts = {}) {
        super();

        const options = this.options = Object.assign({}, this.constructor.Defaults, opts);

        this.value = null;
        this.valueRaw = null;
        this.valueInitial = null;
        this.valueTarget = null;

        this.hasInitialValueSet = false;
        this.isRunning = false;
        this.time = null;
        this.timer = null;

        this.stepBinded = this.step.bind(this);

        if (options.step) {
            this.on('step', options.step);
        }

        if (options.start) {
            this.on('start', options.start);
        }

        if (options.stop) {
            this.on('stop', options.stop);
        }

        if (options.value !== null) {
            this.to(options.value);
        }
    }

    /**
     * Destructor
     */
    destroy () {
        this.listeners = this.options = {};

        if (this.timer) {
            cancelAnimationFrame(this.timer);
        }
    }

    /**
     * Set target value
     *
     * @param {any} valueCurrent
     *
     * @memberOf EaseValue
     */
    to (value) {
        if (!this.hasInitialValueSet) {
            this.reset(value);
        } else {
            this.valueInitial = this.value;
            this.valueTarget = value;

            if (!this.isRunning) {
                this.time = timing.now();

                // Trigger 'start' event
                this.trigger('start', this.value);
                this.step();
            }
        }
    }

    /**
     * Reset value / state
     *
     * @param {any} eventName
     * @param {any} callback
     *
     * @memberOf EaseValue
     */
    reset (value) {
        const precision = this.options.precision;

        if (value !== this.valueRaw || value !== this.valueTarget) {
            this.valueRaw = this.valueInitial = this.valueTarget = value;
            this.value = Math.round(value / precision) * precision;
            this.hasInitialValueSet = true;
            this.time = timing.now();

            this.trigger('start', this.value);
            this.trigger('step', this.value);
            this.trigger('stop', this.value);
        }
    }

    /**
     * Perform animation setp
     *
     * @returns
     * @protected
     */
    step () {
        const precision = this.options.precision;
        const easing = EaseValue.easings[this.options.easing];
        const firstRun = !this.isRunning;
        let   running;

        this.isRunning = running = true;

        if (this.hasInitialValueSet) {
            const valueTarget = this.valueTarget;
            const valueLast = this.value;
            const time = timing.now();
            const tdelta = time - this.time;

            // Calculate new value
            const value = easing.call(this, this, tdelta);

            // Animation is considered to be complete when it would be complete in
            // less than 1 step
            const isComplete = Math.abs(valueTarget - value) < precision;

            // Save value
            this.valueRaw = isComplete ? valueTarget : value;
            this.value = Math.round(this.valueRaw / precision) * precision;
            this.time = time;

            // If there was a change or this is the first call then we trigger
            // step event. We want to do it on first call to make sure 'step' is
            // called at least once
            const delta = valueLast - this.value;

            if (delta || firstRun) {
                this.trigger('step', this.value);
            }

            if (isComplete) {
                this.isRunning = running = false;
                this.trigger('stop', this.value);
            }
        }

        if (running) {
            this.timer = requestAnimationFrame(this.stepBinded);
        }
    }

}


class EaseValueMultiple extends Events {
    constructor (easeValues) {
        super();

        this.easeValues = easeValues;
        this.keys = Object.keys(easeValues);
        this.value = this.getValue();
        this.isRunning = this.getIsRunning();
        this.reqStart = this.reqStop = this.reqStep = null;

        this.triggerStart = this.triggerStart.bind(this);
        this.triggerStop = this.triggerStop.bind(this);
        this.triggerStep = this.triggerStep.bind(this);

        this.keys.forEach(name => {
            easeValues[name].on('start', this.handleStart.bind(this));
            easeValues[name].on('stop', this.handleStop.bind(this));
            easeValues[name].on('step', this.handleStep.bind(this));
            if (!this[name]) this[name] = easeValues[name];
        });
    }

    destroy () {
        const easeValues = this.easeValues;

        if (this.reqStart) {
            cancelAnimationFrame(this.reqStart);
        }
        if (this.reqStop) {
            cancelAnimationFrame(this.reqStop);
        }
        if (this.reqStop) {
            cancelAnimationFrame(this.reqStop);
        }

        this.keys.forEach(name => {
            easeValues[name].destroy();
        });

        this.isRunning = false;
        this.easeValues = {};
        this.value = {};
        this.keys = [];
    }

    to (values) {
        const easeValues = this.easeValues;

        this.keys.forEach(name => {
            if (name in values) {
                easeValues[name].to(values[name]);
            }
        });
    }

    reset (values) {
        const easeValues = this.easeValues;

        this.keys.forEach(name => {
            if (name in values) {
                easeValues[name].reset(values[name]);
            }
        });
    }

    getIsRunning () {
        const easeValues = this.easeValues;
        const activeEaseValues  = this.keys.filter(name => easeValues[name].isRunning);
        return activeEaseValues.length > 0;
    }

    getValue () {
        const easeValues = this.easeValues;
        const value = {};

        this.keys.forEach(name => {
            value[name] = easeValues[name].value;
        });

        return value;
    }

    triggerStart () {
        this.reqStart = null;
        this.trigger('start', this.value);
    }

    triggerStop () {
        this.reqStop = null;
        this.trigger('stop', this.value);
    }

    triggerStep () {
        this.reqStep = null;
        this.trigger('step', this.value);
    }

    handleStart () {
        // Trigger only once, first time
        if (!this.isRunning) {
            if (this.reqStart) {
                cancelAnimationFrame(this.reqStart);
            }

            this.value = this.getValue();
            this.isRunning = this.getIsRunning();
            this.reqStart = requestAnimationFrame(this.triggerStart);
        }
    }

    handleStop () {
        this.isRunning = this.getIsRunning();

        if (!this.isRunning) {
            if (this.reqStop) {
                cancelAnimationFrame(this.reqStop);
            }

            this.value = this.getValue();
            this.reqStop = requestAnimationFrame(this.triggerStop);
        }
    }

    handleStep () {
        this.value = this.getValue();

        if (!this.reqStep) {
            this.reqStep = requestAnimationFrame(this.triggerStep);
        }
    }
}

EaseValue.Events = Events;
EaseValue.EaseValueMultiple = EaseValueMultiple;

EaseValue.multiple = function (params) {
    return new EaseValueMultiple(params);
}


EaseValue.defaultForce = 0.1;
EaseValue.defaultPrecision = 0.01;
EaseValue.defaultEasing = 'easeOut';


EaseValue.easings = {
    'easeOut': function (ease, tdelta) {
        const valueDelta = (ease.valueTarget - ease.valueRaw);
        const force = ease.options.force * tdelta / 16;

        if (valueDelta > 0) {
            return Math.min(ease.valueTarget, ease.valueRaw + valueDelta * force);
        } else {
            return Math.max(ease.valueTarget, ease.valueRaw + valueDelta * force);
        }
    },

    'linear': function (ease, tdelta) {
        const valueDelta = (ease.valueTarget - ease.valueRaw);
        const force = ease.options.force * tdelta / 16;

        if (valueDelta > 0) {
            return Math.min(ease.valueTarget, ease.valueRaw + force);
        } else {
            return Math.max(ease.valueTarget, ease.valueRaw - force);
        }
    }
};


module.exports = EaseValue;
